### ISSUE TITLE
Issue processing HDF files offline 

### ISSUE THREAD

**Issue by mnosshi** [lbusett/MODIStsp#85]
New user, managed with lots of help to install `MODIStsp` on mac os x sierra, R 3.4.
Gui loads fine, but when I click on `load options` and select already downloaded HDF files for MOD13C1 (eg.: MOD13C1.A2000209.006.2015147152824.hdf) from my external hard-drive, I get the following errors:
(R:3198): Gtk-WARNING **: Attempting to store changes into `/Users/magednosshi/.local/share/recently-used.xbel', but failed: Failed to create file “/Users/magednosshi/.local/share/recently-used.xbel.FQHL3Y”: No such file or directory

(R:3198): Gtk-WARNING **: Attempting to set the permissions of `/Users/magednosshi/.local/share/recently-used.xbel', but failed: No such file or directory

Is there a step I am missing in configuring the install?
Thanks,
~m

**Comment by lbusett** [comment]
Hi, 

   first of all, the "load options" button is not meant to select previously download images, but a previously save "json" file containing processing options. In practice, you can set all the processing options as you wish, and then save them for later use. 

Concerning thje GTK error, I don't know. Seems like a permission problem. Did you check if that folder exists and you have read/write permissions ?

**Comment by mnosshi** [comment]
This is my first time using JSON. What do I need to do to load my time-series.hdf?
The ".local/share/recently-used.xbel" file does not exist, even when I looked among hidden files in that directory. 
~m

**Comment by lbusett** [comment]
As I wrote yesterday, to use already downloaded images you should: 

"Put all the hdf images on one folder, and set the parameters in the gui as for a normal processing. Set the option regarding the folder where original modis hdfs should be stored to the folder where you have your hdfs and start the processing. MODIStsp will detect the available hdfs and skip the download phase."

I'd suggest you to read carefully the package vignette and start a "test run" to download just a cuple of dates and examine the output folders to understand what's happening.

**Comment by mnosshi** [comment]
I did read the vignette carefully, and here is all it instructs:
"offline: this option allows to process/reprocess HDF files already
available on the user’s PC without downloading from NASA – useful if the
user already has an archive of HDF images, or to reprocess data already
downloaded via MODIStsp (for example, to create time series for an
additional layer). " there are no additional instructions for selecting the
hdf files under loadoptions, nor how to proceed with the following prompt:
"the selected file is not JSON, do you want to proceed?"
or how to select multiple hdf files.
In the end I have an obscure error message in R:
(R:4001): Gtk-WARNING **: Attempting to set the permissions of
`/Users/magednosshi/.local/share/recently-used.xbel', but failed: No such
file or directory
[Wed Jul 19 10:13:15 2017] Retrieving list of files from NASA server
[Wed Jul 19 10:13:15 2017] No available data for year: 2000 for Sensor
Terra in selected dates.
[Wed Jul 19 10:13:15 2017] No available data for year: 2001 for Sensor
Terra in selected dates.
[Wed Jul 19 10:13:15 2017] No available data for year: 2002 for Sensor
Terra in selected dates.
[Wed Jul 19 10:13:15 2017] No available data for year: 2003 for Sensor
Terra in selected dates.
[Wed Jul 19 10:13:15 2017] No available data for year: 2004 for Sensor
Terra in selected dates.
[Wed Jul 19 10:13:15 2017] No available data for year: 2005 for Sensor
Terra in selected dates.

and no files processed.
What can you suggest?
Thanks,
~m


 and followed your instructions from yesterday. All I get is an empty
folder and the GTK error I mentioned.

**Comment by ggranga** [comment]
If you have read carefully the vignette, you would find also the part which describes how to specify the folder for HDF files:
> ## Output Folder for Original HDF Storage
> Select the folder where downloaded **original MODIS HDF files** will be stored. [...]

Instead, "there are no additional instructions for selecting the hdf files under loadoptions" simply because - I cite what @lbusett wrote you - **"the "load options" button is not meant to select previously download images, but a previously save "json" file containing processing options"**.
This is also documented in the vignette:
> All processing parameters can be easily set with a user-friendly GUI, although non-interactive execution exploiting a previously created Options File is possible. [...]

So, to load your HDF files simply follow the instructions which @lbusett already gave you two times:
1) "Put all the hdf images on one folder"
2) "Set the option regarding the folder where original modis hdfs should be stored to the folder where you have your hdfs" -> not with the "Load Options" button, but **with the "Output Folder for Original HDF files download" option in the GUI** (see the image below)
![bitmap](https://user-images.githubusercontent.com/6874239/28378024-08516cee-6caf-11e7-8429-05630d87f6a3.png)


In this way the HDF folder will be recognised, and so you should be able to process your HDF files.

Regarding the message "the selected file is not JSON, do you want to proceed?": MODIStsp saves the option file in the JSON format, so you receive an alert if you try to load a file that is not a JSON. The vignette does not explain how to proceed with this warning because it is self-explaining.
If you want to understand how the Option File works, try to save an option files with the button "Save Options"; then change some options in the GUI, and after that load the above created JSON options file: you yill see that the options in the GUI will be restored to the saved values. Moreover. since the JSON is a plain text, you can open it with a text editor to explore it.

Finally, the warning is related to a GTK functionality that I do not know, so I cannot help you with it. However, since it is only a warning and not an error, I suggest to ignore it for now (it does not appear to be related with your problem).


This is the way

**Comment by mnosshi** [comment]
While I appreciate your efforts in putting together this package, and
follow up to trouble shoot, which I am happy to do, there is no need for
snide.

I did try all of the above, including steps 1 and 2 you re-listed above.
The best thing I am able to communicate back to you, if there is still any
interest in troubleshooting, is the error message I am receiving, and the
fact that the output file, in this case "VI_16Days_005dg_v6" comes out
completely empty, with no .tiff files.

Again, I was happy to invest the time it took to get this package installed
(thanks to everyone's hlep!), and happy to invest more time to resolve
issues in a friendly manner.

Best,
maged

**Comment by ggranga** [comment]
I am sorry if my answer sound you snide, which was not my intention; nevertheless, if you answered that there are no instructions for selecting hdf files under loadoptions after we specified that the "load options" button is not meant to select hdf files, it was evident that the suggestion was disattended, so I felt obliged to remark that. I would like to help you to solve your problem, but I can do it only if communication will be reciprocal. Hoping this clarification can close this misunderstanding.

The offline mode works in this way: the HDF files which are required for processing (following the criteria passed through the GUI or a JSON Options File) are searched only in the "output folder for HDF files download"; so, if you receive an empty folder, the reason is that no HDF files matching your criteria are found. This can be due to several reasons:
1) the "output folder for HDF files download" is not the folder in which HDF files are stored; 
2) HDF files are stored in subdirectories instead of in the main directory;
3) HDF files do not match the criteria specified with the GUI (i.e. MODIS tiles are different, or HDF files relate to dates outside the specified starting - ending date period);
4) HDF files were renamed.

Please check the above four points; if all of that are attended, to help us to understand what is wrong please save your JSON Options Files and attach it to this issue, together with the list of the content of the HDF folder (also a screenshot is ok).

**Comment by lbusett** [comment]
Hi, 

   I just updated MODIStsp FAQ with some more detailed information concerning processing of already-downloaded hdf files. I hope it will clarify things a bit:

[https://github.com/lbusett/MODIStsp/wiki#old-pdfs)](https://github.com/lbusett/MODIStsp/wiki#old-pdfs)

HTH !

**Comment by mnosshi** [comment]
Thank you for your note, and no worries. It can be easy with limited communications to get a little out of hand. I also was not clear that I did try suggestions by @lbusett with the "Output folder".

OK..
All HDF files are in a single folder and not in subdirectories (points #1 & 2 above). Also, I did not rename files (point #4), which leaves point #3. 
I entered the bounding box coordinates as in the original selection (it might be off slightly, since I did not note the specific coordinates when I made my original selection in EarthData Search when downloading HDF files manually). Here is what I got when I processed the data, attached the following:

1. Selection in GUI
2. json file

No files were processed in the end.
I already went ahead and used the GUI to download and process the data I need and it worked great. If you have any suggestions on why I am not doing this step correct, or if it is in fact because of the bounding box selection coordinates not exactly matching the HDF files, please let me know.

Thanks again for all the help!
~maged
<img width="945" alt="screen shot 2017-07-20 at 10 07 10 am" src="https://user-images.githubusercontent.com/29981357/28425163-192f2e62-6d35-11e7-9270-df2c555496e0.png">

<img width="638" alt="screen shot 2017-07-20 at 10 21 32 am" src="https://user-images.githubusercontent.com/29981357/28425228-3ff2aede-6d35-11e7-8ec8-7ba65336f835.png">

**Comment by lbusett** [comment]
OK. This is much more clear, thanks !  ;-) 

1) Quick thing to be sure : did you press the "Retrieve tiles from bounding box" button to set the tiles extent AND do the tiles shown in the "Required MODIS tiles" window reflect the ones of your hdfs ? 

2) Could you share also a screenshot of your "hdf images" folder showing a list of files ? That way we could check if the regular expression used to identify files would work.

**Comment by mnosshi** [comment]
To answer 1, I did press "Retrieve tiles...", but did NOT enter anything into "Required MODIS Tiles", the box with "Horizontal:Start" etc. stayed greyed out. 
I then entered coordinates for the "Output Bouding Box..." as close to the manual selection in EarthData as I could remember, but they likely do not match exactly since I did not take not of the exact coordinates when I did the manual download. 
Does that matter? If so, this might be the problem, and I'll make note of my selections in the future.

Here is folder with HDF files:

<img width="940" alt="screen shot 2017-07-20 at 10 55 23 am" src="https://user-images.githubusercontent.com/29981357/28427098-ba3e3eb0-6d3a-11e7-88ad-01ea4dfd82ca.png">

**Comment by lbusett** [comment]
Ok. Now the issue is clear: the naming convention of the files downloaded from "EarthData" is completely different from the one used in LPDAAC. MODIStsp is therefore unable to recognize the already available files. 

You can see for example that no indication concerning the "tiles" is available in your files, while a typical "LPDAAC" file would look like: 

"MYD13C1.A2000305.**h18v04**.006.2015139075710.hdf"

In addition, I understand now that the files that you have are not the "original" MODIS hdf files, but subsets that were produced by earthdata. 

So, in short: MODIStsp will **not work** on those files, sorry. You'll have to re-download ! 

The offline functionality is meant to work **only** with the original MODIS hdfs downloaded from LPDAAC (or elsewhere - the important thing is that they are not modified/preprocessed in any way). 
I don't know if we will ever implement functionality for earthdata hdfs (I don't think so, at the moment)

**Comment by mnosshi** [comment]
The MODIStsp download/processing was seamless. Thank you all for that!
I also appreciate the support in resolving issues with installation and processing. 

What would be a good place to post my notes for installation on mac OSX? Otherwise, I am happy to pass them on to others running into the same problems when you come across that.
All the best,
~maged

**Comment by lbusett** [comment]
If you agree, I would be happy to add those in the FAQ of modistsp. I think it would be the best place to put them.

**Comment by mnosshi** [comment]
that sounds fine, I will get my notes together and send those notes this
weekend.

~maged

**Comment by mnosshi** [comment]
Hi,

Here are my notes for installing MODIStsp on a mac (attached)

Best,
Maged

**Comment by lbusett** [comment]
Hi Maged,

   thanks ! However, I didn't receive the attach, sorry. Could you send it to me by mail ? 

By the way: you may want to have a look at the new "MODIStsp" website: 

[http://lbusett.github.io/MODIStsp/](http://lbusett.github.io/MODIStsp/)
 
I think it provides a much better "interface" to the package. I think I will place also your instructions for MAC instasllation there, in the future.

**Comment by mnosshi** [comment]
I have a question about how you create the and save the Raster stack as
RData..
When I create the stack:
> s <- stack(dir())
> save(s,file="~/Documents/PUBS/EVI RDATA/s.Rdata")

then look at the attributes
> s
I get everything as I would with the RData file created by MODIStsp, but
*not* the "time", for example:
time        : 2000-07-27 - 2017-06-26 (range)

How can I add this when creating the Rasterstack?

Thanks
maged

**Comment by lbusett** [comment]
Hi. 

   Note that I moved this topic to a new open issue. I will reply you probably tomorrow.
